tests/rptest/tests/e2e_shadow_indexing_test.py

```diff
@@ -110,3 +123,87 @@ def test_write(self):
+    def check_s3():
+        objects = list(redpanda.get_objects_from_si())
+        return len(objects) > 0
+
+    wait_until(check_s3,
+               timeout_sec=30,
+               backoff_sec=1,
+               err_msg='Failed to write objects to S3')
+
+
+class ShadowIndexingInfiniteRetentionTest(PreallocNodesTest):
+    # Test infinite retention while SI is enabled.
+    # This class ports Test5 from https://github.com/redpanda-data/redpanda/issues/3572
+    # into ducktape.
+    segment_size = 2**30
+    topics = [TopicSpec(partition_count=5)]
+
+    def __init__(self, test_context):
+        self._si_settings = SISettings(
+            log_segment_size=self.segment_size,
+            cloud_storage_cache_size=20 * 2**30,
+            cloud_storage_reconciliation_interval_ms=500,
+            cloud_storage_max_connections=5,
+            cloud_storage_enable_remote_read=False,
+            cloud_storage_enable_remote_write=False)
+
+        super(ShadowIndexingInfiniteRetentionTest,
+              self).__init__(test_context=test_context,
+                             node_prealloc_count=1,
+                             num_brokers=3,
+                             si_settings=self._si_settings)
+
+    @cluster(num_nodes=4, log_allow_list=KGO_ALLOW_LOGS)
+    def test_infinite_retention(self):
+        self.logger.info(f"Environment: {os.environ}")
+        if os.environ.get('BUILD_TYPE', None) == 'debug':
+            self.logger.info(
+                "Skipping test in debug mode (requires release build)")
+            return
+
+        await_bucket_creation(self.redpanda,
+                              self._si_settings.cloud_storage_bucket)
+
+        # Remote write/read and retention set at topic level
+        rpk = RpkTool(self.redpanda)
+        rpk.alter_topic_config(self.topic, 'redpanda.remote.write', 'true')
+        rpk.alter_topic_config(self.topic, 'redpanda.remote.read', 'true')
+        rpk.alter_topic_config(self.topic, 'retention.bytes', '-1')
+        rpk.alter_topic_config(self.topic, 'retention.ms', '-1')
+
+        # 100k messages of size 2**18
+        # is ~24GB of data.
+        msg_size = 2**18
+        msg_count = 100000
+
+        producer = FranzGoVerifiableProducer(self.test_context, self.redpanda,
+                                             self.topic, msg_size, msg_count,
+                                             self.preallocated_nodes)
+        producer.start(clean=False)
+        await_minimum_produced_records(self.redpanda,
+                                       producer,
+                                       min_acked=msg_count / 10)
+        await_s3_objects(self.redpanda)
+
+        rand_consumer = FranzGoVerifiableRandomConsumer(
+            self.test_context, self.redpanda, self.topic, msg_size, 100, 10,
+            self.preallocated_nodes)
+        rand_consumer.start(clean=False)
+
+        producer.wait()
+        rand_consumer.shutdown()
+        rand_consumer.wait()
```

Comment:
Would be good to also run a sequential consumer at end of test, so that we're truly validating that all the messages landed where we expected (random consumer is only going to hit a relatively small subset of messages)